VMware: New module vmware_host_sriov

victron opened the pull request:
##### SUMMARY
Add a new module for changing SR-IOV settings on host interface

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
vmware_host_sriov

##### ADDITIONAL INFORMATION
- This module can be used to configure, enable/disable SR-IOV functions an ESXi host.
- module didn't reboot host after changes, but put in output "rebootRequired" key from vCenter status
- tested on VMware 6.0

victron committed:
vmware_host_sriov: working version

- integration tests (tested on ESXi 6.0)
- unit tests

victron committed:
added some corections for sanity test

victron committed:
corrected sanity errors

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,332 @@
+- This module can be used to configure, enable/disable SR-IOV functions an ESXi host.
+- module didn't reboot host after changes, but put in output "rebootRequired" key.
+- User can specify an ESXi hostname or Cluster name. In case of cluster name, all ESXi hosts are updated.
+version_added: 2.9
```
```suggestion
version_added: 2.10
```

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,332 @@
+notes:
+- Tested on vSphere 6.0
+requirements:
+- python >= 2.6
```
```suggestion
- python >= 2.7
```

Pyvmomi requires 2.7

Akasurde replied:
I know I need to change for the rest of them.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,332 @@
+  vmnic:
+    description:
+    - interface name, like vmnic0
+    type: str
```
Can this be a list of NICs ?

victron replied:
I thought about it. But in reason that it's my first experience, I prefer leave it as is.  A little bit late, I'll redo it on [str,  list].

victron replied:
I'll do it in next version. Because may be it's better use dict instead, for example:
```
[{ "vmnic": "vmnic0",  "num_virt_func": 8},
{ "vmnic": "vmnic1",  "num_virt_func": 0}]

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,332 @@
+    - interface name, like vmnic0
+    type: str
+    required: True
+  sriovOn:
```
```suggestion
  sriov_on:
```

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,332 @@
+    required: True
+  sriovOn:
+    description:
+    - desired SR-IOV state on interface
```
```suggestion
    - Desired SR-IOV state on interface.
```

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,332 @@
+    - desired SR-IOV state on interface
+    type: bool
+    required: True
+  numVirtFunc:
```
```suggestion
  num_virt_func:
```

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,332 @@
+    required: True
+  numVirtFunc:
+    description:
+    - number of functions to activate on interface
```
```suggestion
    - number of functions to activate on interface.
```

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,332 @@
+        )
+        if not self.hosts:
+            self.module.fail_json(msg="Failed to find host system.")
+        # self.change_list = []
```
```suggestion

```

Akasurde reviewed (changes requested):
Overall looks good. @goneri  @pgbidkar @anusha94 @Tomorrow9 Can you please review this ? Thanks.

victron committed:
change ver. to 2.10, corected some lint errors

victron committed:
Update lib/ansible/modules/cloud/vmware/vmware_host_sriov.py

Co-Authored-By: Abhijeet Kasurde <akasurde@redhat.com>

victron committed:
Update lib/ansible/modules/cloud/vmware/vmware_host_sriov.py

Co-Authored-By: Abhijeet Kasurde <akasurde@redhat.com>

victron committed:
Update lib/ansible/modules/cloud/vmware/vmware_host_sriov.py

Co-Authored-By: Abhijeet Kasurde <akasurde@redhat.com>

victron committed:
Update lib/ansible/modules/cloud/vmware/vmware_host_sriov.py

Co-Authored-By: Abhijeet Kasurde <akasurde@redhat.com>

victron committed:
Update lib/ansible/modules/cloud/vmware/vmware_host_sriov.py

Co-Authored-By: Abhijeet Kasurde <akasurde@redhat.com>

victron committed:
Update lib/ansible/modules/cloud/vmware/vmware_host_sriov.py

Co-Authored-By: Abhijeet Kasurde <akasurde@redhat.com>

victron committed:
renaming sriovOn-> sriov_on, numVirtFunc-> num_virt_func

victron committed:
unit tests for python >= 3.7

victron committed:
changing version_added to str

victron committed:
changed on build in units.compat for mocking

pgbidkar commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,331 @@
+module: vmware_host_sriov
+short_description: Manage SR-IOV settings on host
+description:
+- This module can be used to configure, enable/disable SR-IOV functions an ESXi host.
```
```suggestion
- This module can be used to configure, enable/disable SR-IOV functions on ESXi host.
```

pgbidkar commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,331 @@
+short_description: Manage SR-IOV settings on host
+description:
+- This module can be used to configure, enable/disable SR-IOV functions an ESXi host.
+- module didn't reboot host after changes, but put in output "rebootRequired" key.
```
Please can you rephrase this sentence? Something lile
```suggestion
- Module doesn't  reboot the host after changes, but puts it in output "rebootRequired" state.
```

pgbidkar commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,331 @@
+  esxi_hostname:
+    description:
+    - Name of the host system to work with.
+    - This parameter is required if C(cluster_name) is not specified.
```
I think user can specify specific host from cluster. Cluster and host both arguments should be allowed

victron replied:
yes, it allowed.  I added this comment in doc.
`- user can specify specific host from cluster.`

pgbidkar commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,331 @@
+  cluster_name:
+    description:
+    - Name of the cluster from which all host systems will be used.
+    - This parameter is required if C(esxi_hostname) is not specified.
```
I think user can specify specific host from cluster. Cluster and host both arguments should be allowed

victron replied:
yes, it allowed. I added this comment in doc.
`- user can specify specific host from cluster.`

pgbidkar commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,331 @@
+    description:
+    - Desired SR-IOV state on interface.
+    type: bool
+    required: True
```
is default value not required here?

victron replied:
as I understood it done especially:
`./ansible/lib/ansible/module_utils/common/parameters.py`
```
    for (k, v) in argument_spec.items():
        legal_inputs.append(k)
        aliases = v.get('aliases', None)
        default = v.get('default', None)
        required = v.get('required', False)
        if default is not None and required:
            # not alias specific but this is a good place to check this
            raise ValueError("internal error: required and default are mutually exclusive for %s" % k)
        if aliases is None:
```
 I got this error
```
{"failed": true, "msg": "Module alias error: internal error: required and default are mutually exclusive for num_virt_func"}
```

pgbidkar commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,331 @@
+        """Check ESXi host configuration"""
+        change_list = []
+        changed = False
+        for host in self.hosts:
```
is this loop applicable only Cluster name is passed right?

victron replied:
I got this error
```
{"failed": true, "msg": "Module alias error: internal error: required and default are mutually exclusive for num_virt_func"}
```

pgbidkar replied:
I didnt get this. Did you make any changes to loop and got this error?

victron replied:
sorry, ignore my previous post - wrong thread.
> is this loop applicable only Cluster name is passed right?

yes, you are right

pgbidkar commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,331 @@
+    argument_spec = vmware_argument_spec()
+    argument_spec.update(
+        cluster_name=dict(type="str", required=False),
+        esxi_hostname=dict(type="str", required=False),
```
cluster_name or esxi_hostname any one of them should be required.

victron replied:
yes, but not both of them in a same time.
this check done in
```
module = AnsibleModule(
        argument_spec=argument_spec,
        required_one_of=[["cluster_name", "esxi_hostname"]],
```
in
`./ansible/lib/ansible/module_utils/common/validation.py`
```
def check_required_one_of(terms, module_parameters):
    """Check each list of terms to ensure at least one exists in the given module
    parameters

    Accepts a list of lists or tuples
```

pgbidkar replied:
What I meant is one of them should be made required=True. 
What happens if user doesnt pass both of them?

victron replied:
> What happens if user doesnt pass both of them?

all will be fine
```
vic@AN: ~/ansible vmware_host_sriov
$ cat ./t_args.json                                                                                      [14:25:08]
{
        "ANSIBLE_MODULE_ARGS": {
                "hostname": "10.72.24.143",
                "username": "administrator@vsphere.local", 
                "password": "Aff", 
                "validate_certs": false, 
                "sriov_on": true,
                "num_virt_func": 8, 
                "vmnic": "vmnic7"
        }
}

(envPy37_ansible_dev)  
vic@AN: ~/ansible vmware_host_sriov
$ python -m ansible.modules.cloud.vmware.vmware_host_sriov ./t_args.json                                 [14:25:30]

{"msg": "one of the following is required: cluster_name, esxi_hostname", "failed": true, "invocation": {"module_args": {"hostname": "10.72.24.143", "username": "administrator@vsphere.local", "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "validate_certs": false, "sriov_on": true, "num_virt_func": 8, "vmnic": "vmnic7", "port": 443}}}
(envPy37_ansible_dev) FAIL: 1

vic@AN: ~/ansible vmware_host_sriov
```

pgbidkar commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,331 @@
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        required_one_of=[["cluster_name", "esxi_hostname"], ["sriov_on", "num_virt_func"]],
+        supports_check_mode=True,
```
I think this not handled?

victron replied:
it, handled. If to look into
`./ansible/lib/ansible/module_utils/common/validation.py`
```
def check_required_one_of(terms, module_parameters):
    """Check each list of terms to ensure at least one exists in the given module
    parameters

    Accepts a list of lists or tuples
```
but now it doesn't matter, I simplified  this part. 
```
    module = AnsibleModule(
        argument_spec=argument_spec,
        required_one_of=[["cluster_name", "esxi_hostname"]],
        supports_check_mode=True,
    )
```

pgbidkar commented on test/integration/targets/vmware_host_sriov/tasks/main.yml:
```diff
@@ -0,0 +1,102 @@
+# Test code for the  module vmware_host_sriov
+
+- when: vcsim is not defined
```
This block will skip all tests in integration test. @goneri please can you confirm if this block is required?

goneri replied:
The question is more: Can we run the integration test-suite with govcsim? if it's the case, we should drop the line. Otherwise, the integration tests will still be run, but only with a regular lab.

victron replied:
as I understood there are no possibility to test it with `govcsim`
```
$ curl -sk https://user:pass@127.0.0.1:8989/about | grep -i configManager  
FAIL: 1
```
May be I am wrong,  please point me.  

> if it's the case, we should drop the line. therwise, the integration tests will still be run, but only with a regular lab.

I thought, that `- when: vcsim is not defined` - instructs not to use `vcsim` and forward directly to regular lab tests. It's not correct?

anusha94 commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,331 @@
+    description:
+    - number of functions to activate on interface.
+    - if sriov_on false should be equal 0
+    - if sriov_on true should be more then 0
```
if sriov_on true should be more than 0

anusha94 reviewed (commented):
Since there are scenarios like 
`either one of esxi_host or cluster is required`, 
`if sriov_on is false, then num_virt_func = 0`, these can be checked for in the beginning itself and you can `exit_json` with appropriate messages

IMO, it is better to have a function to sanitize params like in [vmware_guest_network](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/vmware/vmware_guest_network.py#L331)

victron committed:
Update lib/ansible/modules/cloud/vmware/vmware_host_sriov.py

Co-Authored-By: Pavan Bidkar <pbidkar@vmware.com>

victron committed:
Update lib/ansible/modules/cloud/vmware/vmware_host_sriov.py

Co-Authored-By: Pavan Bidkar <pbidkar@vmware.com>

victron committed:
added sanitize_params function

doing user input check and system compatibility

victron committed:
sriov_on required=False

set sriov_on as optional parameter, all settings can be done via num_virt_func

victron commented:
> 
> 
> Since there are scenarios like
> `either one of esxi_host or cluster is required`,
> `if sriov_on is false, then num_virt_func = 0`, these can be checked for in the beginning itself and you can `exit_json` with appropriate messages
> 
> IMO, it is better to have a function to sanitize params like in [vmware_guest_network](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/vmware/vmware_guest_network.py#L331)

thanks @anusha94 , thank you for advice I'll do it.

victron committed:
tests for sanitize_params method

removed old test due to new mmethod

victron committed:
FIX: fixes for sanitize_params method

victron commented:
main changes in coomit
```
num_virt_func:
    description:
    - number of functions to activate on interface.
    - 0 means SR-IOV disabled.
    - number greater than 0 means SR-IOV enabled.
    type: int
    required: True
  sriov_on:
    description:
    - optional parameter, related to num_virt_func.
    - SR-IOV can be enabled only if num_virt_func > 0.
    type: bool
    required: False
```
and added  function to sanitize params.
Please make a review.

victron committed:
FIX: linter errors

victron committed:
comments added to clarify how hosts selected and now changes applied for them

victron committed:
removed trailing-whitespace

anusha94 commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,387 @@
+  sriov_on:
+    description:
+    - optional parameter, related to num_virt_func.
+    - SR-IOV can be enabled only if num_virt_func > 0.
```
SR-IOV can be enabled only if `C(num_virt_func) > 0.`

victron replied:
ok, thanks

anusha94 commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,387 @@
+    required: True
+  sriov_on:
+    description:
+    - optional parameter, related to num_virt_func.
```
optional parameter, related to `C(num_virt_func)`.

victron replied:
ok, thanks

anusha94 commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,387 @@
+            self.module.fail_json(msg="Failed to find host system.")
+        self.results = {"before": {}, "after": {}, "changes": {}}
+
+    def sanitize_params(self, before, hostname, vmnic):
```
I am not clear as to why `before` `hostname` and `vmnic` are required, whatever the input parameters can be obtained from `module.params`

victron replied:
ok, now I got your point. I'll redo.

anusha94 commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,387 @@
+            )
+
+            # check compatability
+            if not before["sriovCapable"]:
```
checking compatibility can be a different method. It doesn't have to be under `sanitize_params`.
Or change the method name to something more appropriate

victron replied:
thank you, I'll do separate method.

anusha94 commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,387 @@
+                "before"
+            ][host.name]["rebootRequired"]
+
+            params_to_change = self.sanitize_params(
```
If you choose to make `sanitize_params` and `check_compatibility` separate methods, `sanitize_params` can be called even before the `for loop` above

victron replied:
thanks, it will look more clear.

anusha94 commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -0,0 +1,387 @@
+                ]
+                params_to_change["change"] = True
+            else:
+                params_to_change["changes"]["msg"] = "Not active (looks not rebooted) "
```
Not active (looks **like** not rebooted)?

victron replied:
ok

victron committed:
corrected some typos

thanks to @anusha94

victron committed:
FIX: integration test: corected test description

victron committed:
TEST: new cases

- test_sanitize_params
- test_check_compatibility
- test_make_diff

victron committed:
any modify of self.results moved out from _update_sriov

victron committed:
sanitize_params fragmented

- sanitize_params: checks user input and just raise error if any
- check_compatibility: checks hardware compatibility with user input, raise error if any
- make_diff: preparing changes list which will be applied
updated set_host_state for new methods

anusha94 commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -184,117 +184,112 @@ def __init__(self, module):
                     msg="with sriov_on == true,  alowed value for num_virt_func > 0"
                 )
-            params_to_change.update({"sriovEnabled": False, "numVirtualFunction": 0})
+            self.sriov_on = False  # fill value, if urser not provided
```
typo for `user`

anusha94 replied:
instead of setting to `False` here, can you make it as a default value if not provided?

victron replied:
I think not. Because I need give two values into `nic_sriov = vim.host.SriovConfig()`.
PyVmomi (VMware) requires two parameters. So, at this point it's better to set correct parameter.

anusha94 replied:
okay, seems right.
Just the typo correction for `user` is required.
Everything else looks good :)

victron replied:
I did.
Anusha, thank you very much for your advices. Now it looks better )

anusha94 commented on lib/ansible/modules/cloud/vmware/vmware_host_sriov.py:
```diff
@@ -184,117 +184,112 @@ def __init__(self, module):
-            params_to_change.update(
-                {"sriovEnabled": True, "numVirtualFunction": self.num_virt_func}
-            )
+            self.sriov_on = True  # fill value, if urser not provided
```
same as above, since `sriov_on` is an optional arg, looks like it can be handled with assigning a default value.

anusha94 reviewed (commented):
Just a few minor observations. Please look into it.

victron committed:
typo correction

anusha94 reviewed (approved):
shipit

goneri commented:
Thanks @victron for the patch and @anusha94 for the review. I don't have the hardware to test the module, but it looks really good to me :-).

Akasurde commented:
Superseded by https://github.com/ansible-collections/vmware/pull/267
